### ISSUE TITLE
`Eq` implementation of `WatchDescriptor` is wrong

### ISSUE THREAD

**Issue by hannobraun** [hannobraun/inotify-rs#60]
From [Stack Overflow](https://stackoverflow.com/questions/45315521/how-do-i-use-the-watchdescriptor-struct-generated-by-inotify-rs-to-match-events/45316225#45316225):

> As noted above, inotify watch descriptor values are scoped per parent inotify file descriptor. This makes the use of [derive(Eq)] on WatchDescriptor highly questionable. Looks like a straight-up bug.

I haven't had the chance to look into the details myself so far, but this seems legit.

**Comment by arsalan86** [comment]
I have tested this by creating multiple `Inotify` instances and matching the `WatchDescriptors` they generate.

`assert_eq!(wd_from_inotify1, wd_from_inotify2);` should not be possible.

While the `Eq` trait does seem useful within the scope of an `Inotify` instance, the potential to match across instances can break things.

I don't think removing the `Eq` trait is the right way to address this issue. Perhaps it makes more sense for the `Inotify` struct to generate an identifier unique to itself. This can be passed to generated `WatchDescriptor`s as a field. This would retain the usability of the `Eq` trait within the scope of the parent `Inotify`, and protect against matching `WatchDescriptor`s between instances of `Inotify`

**Comment by hannobraun** [comment]
I think the most straight-forward way to address this would be to add the inotify instance's file descriptor to `WatchDescriptor`. That would do nothing to solve #61, but at least the `Eq` implementation would be working correctly.

**Comment by arsalan86** [comment]
Yes, that does seem to be the most practical way of fixing the `Eq` implementation. I've implemented a field of type `RawFd` in the `WatchDescriptor` struct in my working tree.

I'll do a PR once I've figured out what can be done about #61. My tree doesn't build at the moment.

**Comment by hannobraun** [comment]
If you make a pull request for adding the `RawFd', I'd definitely accept it. Yes, eventually we want a complete solution that also solves #61, but I suspect that will require more thinking and therefore more time. Until then, fixing the `Eq` issue in a simple way will be very welcome.

**Comment by arsalan86** [comment]
Adding the `RawFd` field to `WatchDescriptor` introduces another issue: Inotify itself doesn't associate watch descriptors and events with a particular Inotify instance.

From the Inotify source (accessed from [here](http://elixir.free-electrons.com/linux/latest/source/fs/notify/inotify/inotify_user.c#L189)):

	pad_name_len = round_event_name_len(fsn_event);
	inotify_event.len = pad_name_len;
	inotify_event.mask = inotify_mask_to_arg(fsn_event->mask);
	inotify_event.wd = event->wd;
	inotify_event.cookie = event->sync_cookie;

Implementing a mechanism to prevent `WatchDescriptors` from being shared between `inotify-rs` instances is possible, but it requires changes to `inotify-rs` such that the `Inotify` and `Event` by adding a `RawFd` field to the `WatchDescriptor` struct. The `add_watch` method would then need to return a `WatchDescriptor` with a RawFd field assigned `self.0`:

    match wd {
        -1 => Err(io::Error::last_os_error()),
        _  => Ok(WatchDescriptor(wd, self.0)),
    }

This adds extraneous information to the `WatchDescriptor` returned by inotify-sys, and consistency checks at the `rm_watch` function and during event generation will need to be handled internally by `inotify-rs`.

Moreover, since Inotify doesn't include information about the originating Inotify instance with watch descriptors, inotify will need to implement this at event creation.

This is a major change that affects how the library handles events and watch descriptors, and would also be a breaking change.

**Comment by hannobraun** [comment]
> This is a major change that affects how the library handles events and watch descriptors, and would also be a breaking change.

I think the additional overhead on top of inotify is warranted in this case. I believe it is reasonable to expect an idiomatic Rust wrapper to prevent misuse of the underlying API, as far as that is practical.

Breaking changes are acceptable at this stage. That's why the library is still at an 0.x version.

**Comment by arsalan86** [comment]
If there’s going to be breaking changes, why not use the opportunity to provide an easier and more manageable interface to the user?

I think making use of a hashmap for indexing watch descriptors is a good way forward:

The `add_watch()` method creates a HashMap<key, watchdescriptor> and returns the key.

The `rm_watch()` method takes this key and looks it up in the HashMap.

Events are also returned with the same key to link them to HashMaps.

The keys could be unique (how?) to prevent contamination between instances of `Inotify`. Alternatively, the HashMap itself could be passed to the user as Inotify.HashMap.

I’m just thinking aloud right now. I’m going to try something and push a branch if it seems to work.

**Comment by hannobraun** [comment]
I'm don't understand what the benefit of this would be. Why give the user a key to refer to the `WatchDescriptor` instead of just giving them the `WatchDescriptor` directly? What is the purpose of the `HashMap`?

**Comment by arsalan86** [comment]
I'm sorry about that momentary lapse of reason. Turns out I was doing something very wrong.

I'm making a PR for this bugfix.

**Comment by arsalan86** [comment]
I believe this should be closed.

**Comment by hannobraun** [comment]
I think this should stay open. While #67 addressed the main defect, I thought of another case where the `Eq` implementation could be wrong: If the inotify file descriptor is being reused, and an old `WatchDescriptor` is still around.

Usually, this shouldn't be a problem. You can just create one `Inotify` instance and use that forever. However, people might use the library differently (see #68, where a user talks about one `Inotify` instance per watched directory). A long-running process that watches many directories and opens/closes many files might run into this bug sooner or later.

I have an idea on how to address this (warning, pure idea, not tested in any way): Convert the `fd` in `Inotify` from a `RawFd` to a [`Rc<RawFd>`](https://doc.rust-lang.org/std/rc/struct.Rc.html). Convert `fd` in `WatchDescriptor` to a [`Weak<RawFd>`](https://doc.rust-lang.org/std/rc/struct.Weak.html) that points to the same `RawFd`. Methods that handle `WatchDescriptor` would still need to compare the two `RawFd`s, however, by accessing the `WatchDescriptors`'s `RawFd` through the `Weak` pointer, they would make sure the original `RawFd` was never dropped.

I think this would also solve other issues mentioned in #61. However, it would come with additional cost, most notably one heap allocation per `Inotify` instance. This sounds acceptable to me, but maybe we should ask the [notify](https://crates.io/crates/notify) developers for feedback before merging anything like this. On the other hand, no one ever complained about the heap-allocated buffer in `Inotify` when that was still a thing.

**Comment by hannobraun** [comment]
I did a bit of research. It seems that Linux immediately re-assigns the numbers from closed file descriptors. I've tested with this piece of code:
```rust
let inotify_1 = Inotify::init().unwrap();
print!("{}\n", inotify_1.as_raw_fd());
inotify_1.close().unwrap();
let inotify_2 = Inotify::init().unwrap();
print!("{}\n", inotify_2.as_raw_fd());
```
It prints the same number twice.

This makes it somewhat likely to run into this issue, when handling multiple `Inotify` instances. Knowing that, I think the overhead of the solution I outlined above is clearly worth it (altough I still haven't looked at the solution any closer, so I don't know if it will work out).
